cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,499 @@
+    return *this;
+  }
+
+  StaticVectorImpl& operator=(std::vector<T>&& other) noexcept {
+    assign_by_moving(other.size(), other.data());
+    return *this;
+  }
+
+  // Constructing from count and optional initialization value
+  explicit StaticVectorImpl(size_t count) {
+    storage_.bump_size(count);
+    auto* p = data_ptr();
+    for (size_t i = 0; i < count; ++i) {
+      new (&p[i]) T();
+    }
+  }
+
+  StaticVectorImpl(size_t count, const T& value) {
+    storage_.bump_size(count);
+    auto* p = data_ptr();
+    for (size_t i = 0; i < count; ++i) {
+      new (&p[i]) T(value);
+    }
+  }
+
+  StaticVectorImpl(std::initializer_list<T> values) {
+    storage_.bump_size(values.size());
+    auto* p = data_ptr();
+    for (auto&& v : values) {
+      // XXX cannot move initializer values?
+      new (p++) T(v);
+    }
+  }
+
+  constexpr bool empty() const { return storage_.size_ == 0; }
+
+  constexpr size_t size() const { return storage_.size_; }
+
+  constexpr size_t capacity() const { return storage_.capacity(); }
+
+  constexpr size_t max_size() const { return storage_.max_size(); }
+
+  T& operator[](size_t i) { return data_ptr()[i]; }
+
+  constexpr const T& operator[](size_t i) const { return const_data_ptr()[i]; }
+
+  T& front() { return data_ptr()[0]; }
+
+  constexpr const T& front() const { return const_data_ptr()[0]; }
+
+  T& back() { return data_ptr()[storage_.size_ - 1]; }
+
+  constexpr const T& back() const { return const_data_ptr()[storage_.size_ - 1]; }
+
+  T* data() { return data_ptr(); }
+
+  constexpr const T* data() const { return const_data_ptr(); }
+
+  iterator begin() { return iterator(data_ptr()); }
+
+  constexpr const_iterator begin() const { return const_iterator(const_data_ptr()); }
+
+  iterator end() { return iterator(data_ptr() + storage_.size_); }
+
+  constexpr const_iterator end() const {
+    return const_iterator(const_data_ptr() + storage_.size_);
+  }
+
+  void reserve(size_t n) { storage_.reserve(n); }
+
+  void clear() { storage_.clear(); }
+
+  void push_back(const T& value) {
+    storage_.bump_size(1);
+    new (data_ptr() + storage_.size_ - 1) T(value);
+  }
+
+  void push_back(T&& value) {
+    storage_.bump_size(1);
+    new (data_ptr() + storage_.size_ - 1) T(std::move(value));
+  }
+
+  template <typename... Args>
+  void emplace_back(Args&&... args) {
+    storage_.bump_size(1);
+    new (data_ptr() + storage_.size_ - 1) T(std::forward<Args>(args)...);
+  }
+
+  template <typename InputIt>
+  iterator insert(const_iterator insert_at, InputIt first, InputIt last) {
+    const size_t n = storage_.size_;
+    const size_t it_size = static_cast<size_t>(last - first);  // XXX might be O(n)?
+    const size_t pos = static_cast<size_t>(insert_at - const_data_ptr());
+    storage_.bump_size(it_size);
+    auto* p = data_ptr();
+    if (it_size == 0) {
+      return p + pos;
+    }
+    const size_t end_pos = pos + it_size;
+
+    // Move [pos; n) to [end_pos; end_pos + n - pos)
```

Comment:
In this sort of situation, the STL algos are actually nice
```suggestion
    // Move [pos; n) to [end_pos; end_pos + n - pos)
    std::move(p + pos, p + size(), p + end_pos);
    std::copy(first, last, p + pos);
```

Reply:
Hmm... this doesn't work, because we need to _construct_ the new elements in the array (see the placement new invocations). Which is why, both for moving and for copying, there's a first loop which stops at `n` and another loop which handles the rest.